File: 20-valid-parentheses/valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    # input: a string s
    # Output: boolean true or false
    return false if s.length == 1

    stack = []

    s.chars.each do |char|
        if "({[".include?(char)
            stack << char
            next
        end
    
        if stack.last == "(" && char == ")" ||
            stack.last == "[" && char == "]" ||
            stack.last == "{" && char == "}"
            stack.pop
        else 
            return false
        end
    end

    stack.length == 0
end

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # given a string as input
        # output return a boolean true of false if its a valid string
        stack = []
        for char in s:
            if char == "[" or char == "(" or char == "{":
                stack.append(char)
                continue
            if ((len(stack) == 0) and (char == "]" or char == "}" or char == ")")):
                return False
            else:
                curr_char = stack.pop()
                if ((curr_char == "[" and char == "]") or
                    (curr_char == "{" and char == "}") or
                    (curr_char == "(" and char == ")")):
                    continue
                else:
                    return False
        return len(stack) == 0



        